### ISSUE TITLE
Razor React Project

### ISSUE THREAD

**Issue by TFTomSun** [dsuryd/dotNetify#62]
Hi @dsuryd,

i am fascinated by the project you've created here. My goal is to find the best solution to create SPA with ASP.NET Core. I like the new razor pages, so i've created a first version of a [Razor React](https://github.com/TFTomSun/AspNetCore.RazorReact) project that is based on dotNetify. It allows you to define react components directly in razor pages with type safe data bindings. No jsx files are required. Would be happy if you could have a look at it and tell me what you think about. Maybe we can collaborate. I am very new to web development, so I would be interested in your future plans, what pitfalls need to be solved etc. and what are the most important features that i have to offer to fully support what dotNetify can do.

Btw. i am also new to GitHub. Is there a way to send private messages?

Kind Regards,
Thomas

**Comment by dsuryd** [comment]
Hi, I'm happy with what you're doing there.   This looks good for quickly setting up simple web apps (in terms of UI/UX workflow) without requiring much knowledge of JSX or javascript.  Another value Razor can bring is server-side state hydration, so the page would appear to load faster.  

I'm not sure how far you can go with Razor abstractions on nested components and using existing components made by the React community  - I recommend you refactor my [Composite View example](http://www.dotnetify.net/react/CompositeView) to get an idea.  

My opinion is that JSX (or TSX) in combination with webpack and some popular npm libraries will still give the best developer experience for high-end, component-based, front-end development, and it's really worthwhile for us .NET developers to invest time in learning them ([here's a very good article](https://medium.com/@peterxjang/modern-javascript-explained-for-dinosaurs-f695e9747b70)).  

You can PM me:  dsuryd at gmail.

**Comment by TFTomSun** [comment]
Thanks for the interesting article. I was not aware of the transpile steps. I would have to do that on runtime on server side. I read already a bit about it and it seems to be possible. I'll try to provide a razor based version of your example. But I see that its not so easy as i thought. I've one more question. Is it possible to debug the jsx files in Visual Studio? Or how do you debug the react components?

**Comment by dsuryd** [comment]
I use chrome dev tools window; there’s also a plugin to inspect React components.

**Comment by TFTomSun** [comment]
I made a step forward. I can now define react components as razor view components. I can also nest components into other components. I have one issue at the moment. If I render two instances of the same react component, just the first connects / gets values. The second is rendered but doesn't get any updates from the view model.  Is that a known issue or do you think something is wrong in my implementation?

**Comment by dsuryd** [comment]
There's currently a limitation of one active vm instance of the same type.

**Comment by TFTomSun** [comment]
Are there any plans to be able to decide whether many react classes connect to one single view model instance or to one corresponding view model instance?

**Comment by dsuryd** [comment]
I recommend reading Thinking in React from React’s official website, and Presentational and Container Components article by Dan Abramov.  Associate view models with container components.

The TextBox is bound to props.value, so you need to handle the onChange and update the value. This is called Controlled Form Component (from React website).

**Comment by TFTomSun** [comment]
yes, thanks ... i read a bit more about react and found out that actually the state should be used for internally changing variables. So I have refactored a bit the textbox react class and it works now. I am now able to extract and reuse react classes as razor view components. Its possible to declare them raw inside the view component or in a declarative way. I can also reuse react components inside other react classes. So I think i get closer to be able to use my solution in a real world environment. The Hello Word (client side name computation and the server side clock) example is already implemented and available here:

[Demo App Repository](https://github.com/TFTomSun/AspNetCore.RazorReact.DemoApp)

The syntax is not yet perfect, but I refactored the whole stuff to be very flexible for future requirements. I've already started a way to define jsx functions in C#, that will be transpiled to jsx during the razor processing.
